src/dmdps/docs/dmdps.1
.\" 
.\"									
.\"	Copyright (c) 1987,1988,1989,1990,1991,1992   AT&T		
.\"			All Rights Reserved				
.\"									
.\"	  THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T.		
.\"	    The copyright notice above does not evidence any		
.\"	   actual or intended publication of such source code.		
.\"									
.\" 
'\" t
.fp 2 CW
.lg 0
.TH DMDPS 1 "EXPTOOLS - DMD"
.SH NAME
dmdps \- graphics/text printer support for Teletype 5620 DMD
and 630 MTG terminals 
.SH SYNOPSIS
.I dmdps
{-T printer type}
[\fB-P\fP]
[\fB-7\fP | \fB-8\fP]
[\fB-9600\fP | \fB-4800\fP | \fB-1200\fP]
[\fB-f\fP file name]
[\fB-c\fP pipe command]
.SH DESCRIPTION
.PP
.I Dmdps 
is a layers based program that runs under 
.I jx
to provide general support for text and bitmap output
on 5620 DMD and 630 MTG terminals. 
When 
.I dmdps
is called, it loads itself into the terminal, and it waits for the user to
select a bitmap or for input to be supplied via the 
.I dmdpr
command. 
The available bitmap sources are found on the button 3 menu:
.sp
.TS
center, box;
lF2.
read bitmap
choose layer
whole screen
sweep rectangle
.TE
.in +.5i
.ll -.5i
.sp
Selecting 
.I "read bitmap"
will read the source bitmap from a host file.
You will be prompted for the name of the file.
.sp
Selecting 
.I "choose layer"
will change the mouse icon to the target icon, 
and you will be asked to point at the layer
to print and to click button 3. 
.sp
Selecting 
.I "whole screen"
will select the whole screen. 
.sp
Selecting 
.I "sweep rectangle"
will change the mouse icon to the
box cursor, and you will be asked to sweep a rectangle around
the area to print.
.ll +.5i
.in -.5i
.sp
.PP
In the 
.I dmdps
layer, you can tell what button does what by 
looking at the graphical display running across the top
of the layer. 
The leftmost box is used to indicate the
function of button 1; the middle box, the function of button 2; and the third
box over indicates button 3 function. 
Icons representing button actions/functions appear in these boxes.
The two other boxes send line feeds or form feeds to the printer 
port when button 1 is pressed while being positioned over them. 
The bottom of the 
.I dmdps
layer is a text display line that shows prompts and all typed input. 
The middle portion of the layer shows the selected bitmap.
.PP
When you successfully select a bitmap, it is displayed in the
.I dmdps 
layer.
You can then perform an image operation on the bitmap (bttn3), 
or output the bitmap (bttn2).
Here are the possible contents of the button 2 output and the
button 3 image operations menus:
.sp
.TS
center;
| lF2 | lF2 | lF2 |.
_		_
write bitmap		options...
pipe  bitmap		make a copy
[ print bitmap ]		flip stipple 
		reverse video
		inset rectangle
		[ run / halt ]
_		_
.TE
.in +.5i
.ll -.5i
.sp
The 
.I "print bitmap"
option will format and print the selected bitmap on
the DMD-attached printer.
If the "hex debug dump" flag is turned set,
a hex dump of the formatted output is produced for debugging 
purposes. This will only appear if a printer name was provided
on the command line.
.sp
The 
.I "pipe bitmap"
option will present a shell pipeline which
can be edited by backspacing and typing, or used as is by 
pressing button 3. The pipeline will be fed a run
encoded bitmap image. 
.sp
The 
.I "write bitmap"
operation works just like 
.I "pipe bitmap"
does, except that it writes the encoded bitmap to a host file.
.sp
The 
.I option... 
selection causes the button 2 menu to be temporarily replaced
by an options menu.
.sp
The
.I "make a copy"
operation clones the selected image in off-screen memory, if possible.
.sp
The
.I "flip stipple"
operation flips any stipples in the selected bitmap.
Stipples were used in early versions of the 5620 DMD software
to indicate non-current layers.
.sp
The
.I "reverse video"
operation reverses the video of the selected bitmap.
.sp
Every selection of the
.I "inset rectangle"
operation will mask off four border pixels from
the selected bitmap. 
This is used to make borderless pictures of layers.
.sp 
The
.I halt
operation allows you to halt an active layer before printing.
This item will only appear if you choose a layer other than
the dmdps layer.
.I
The
.I run
operation restores the layer.
.ll +.5i
.in -.5i
.sp
Anything you do to an on-screen bitmap, such as reversing its video
or halting its associated process, will be restored 
after the layers image has been spooled in memory, or when
you select a new bitmap.
.SH OTHER FEATURES
.I dmdps 
reads its host stdin looking for work for the printer -
.I dmdpr 
forwards its own stdin to the xt device of the most
recently loaded 
.I dmdps 
process. 
Thus
.I dmdps 
and 
.I dmdpr 
work together to
form a host pipeline sink. Any kind of data suitable
for the dmd attached printer can be sent down the sink -
including formatted bitmap images, plain ascii text, and
printer-specific escape sequences.
.PP
.I dmdps 
also looks to see if anything is typed at the keyboard
while it it selected. If so, the keyboard input is sent to
the attached printer. This is a easy way to test a printer
connection, to jot down phone numbers, or to send printer escape
codes to change print modes.
.SH OPTIONS
The following command line options are understood by
.I dmdps:
.TP
.BI -P
This tells 
.I dmdps 
to use Kaplan's parallel
centronics port board. This option should be used only if the 
board is attached, otherwise your terminal may crash. 
.sp
When using this board you will want to have your printer
powered off if the 
.I dmdps 
driver is not running. 
Powering on your terminal with the board installed and the 
printer turned on will hang the terminal with a message about
a stray interrupt. Turn off the printer and try again.
.TP
.BI -c " pipe command"
command to be piped to if the 
.I "pipe bitmap" 
menu item is selected. If present, this item must appear last.
.TP
.BI -f " file name"
name of file to put bitmap in if the 
.I "write bitmap" 
menu item is selected.
(Defaults to BLITMAP.)
.TP
.BI -T " printer type"
type of printer to format bitmap for if the 
.I "print bitmap" 
menu item is selected.
The dmdps shell script has built-in printer formatting descriptions (pfd's)
for at least the following printers:
hpink.epson (Thinkjet in Epson mode), 
8510b (C.Itoh 8510b), 
t5310 (Teletype 5310 w/old ROMs), 
t5310aab (Teletype 5310 w/new ROMs), 
rx80 (Epson RX-80), 
or hpink (Thinkjet in native mode.) 
If you present dmdps with a printer name that begins with /, 
dmdps will interpret this as the pfd file name. 
If you ask for the printer type "switches", dmdps will print out the
switch settings for all supported printers.
.TP
.BI -7 | -8 
Selects either a 7 bit byte with odd parity or an 8 bit byte with
even parity for output over the 5620 DMD serial printer port.
(defaults to 7 bits for the T5300 and HPink printers, 
and 8 bits for the others.)
.TP
.BI -9600 | -4800 | -1200
Selects data output rate for the 5620 DMD serial printer port.
(defaults to 9600 bps for the HPink, 4800 bps for the others.)
.SH EXAMPLES
.sp
dmdps -T switches
.PP
The above command will cause dmdps to print out the switch settings
for all supported printers.
.sp
dmdps -P -Thpink
.PP
The above command loads the parallel port version
of the dmdps program, and initializes the formatter 
for HP inkjet output.
.sp 2
dmdps -c "dmdscale -h 3 -w 3 | i300 -l bitmap -d ihXXX"
.PP
The above command loads dmdps and preloads the 
.I "pipe bitmap" 
menu entry to send bitmaps to the Imagen 8/300 printer 
at rje destination ihXXX.
.sp 2
dmdps -c "dmdfmt -p | lpr -b"
.PP
The above command loads dmdps with
.I "pipe bitmap"
preloaded to send bitmaps to a printronix printer.
.sp 2
dmdps -c "dmdfmt -e | dmdpr -u root"
.PP
The above command preloads the "pipe bitmap"
command to format a bitmap for an Epson printer
and forward it to user root's attached DMD printer.
.sp 2
dmdps -T rx80 -f ${HOME}/.bitmap
.PP
The above command will default the print bitmap format to
the Epson RX-80, and will default the output file name to
${HOME}/.bitmap.
.sp 2
dmdpr < file
.PP
The above command will print the file if dmdps is running.
.PP
.I dmdps
runs only under
.IR layers (1).
.SH "FILES"
.ta \w'${HOME}/rje/.prtty    'u
${HOME}/rje	this directory must allow write permission to "all".
.br
${HOME}/rje/.prtty	xt-tty of dmdps process.
.SH "BUGS"
You cannot use the 
.I "pipe bitmap" 
option to send a bitmap to the
host, format it, and send it back to the terminal all in one step.
You can 
.I "write bitmap" 
first, and later format and download it.
The printer port on the 5620 DMD is write-only and cannot
perform any kind of handshaking with the attached printer.
This precludes the ability to do flow-control. This
software attempts to patch around this problem, but does so
at the expense of printing speed. When running the printer
port at 9600 bps, all requests to ring the bell will be sent
to the printer port. If this happens during bitmap output,
the picture will get mangled.
.SH "SEE ALSO"
.IR "dmdfmt(1), dmdscale(1), dmdpr(1), layers(1)."


src/dmdps/src/dmdps.c
/* */
/*									*/
/*	Copyright (c) 1987,1988,1989,1990,1991,1992   AT&T		*/
/*			All Rights Reserved				*/
/*									*/
/*	  THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T.		*/
/*	    The copyright notice above does not evidence any		*/
/*	   actual or intended publication of such source code.		*/
/*									*/
/* */
#include "dmdps.h"
#include "defs.h"
#include "printer.h"
#include "icons.h"
#include "menus.h"
#include "pfd.h"

#define DMDPSVERS "dmdps version 1.2"

/* 
	This code was based on "blitblt", author unknown.
	It has been mostly rewritten by andy schnable.
	The code to drive the parallel port was provided
	by Alan Kaplan in MH, but I don't think he wrote it...	
	The code for the serial port was provided by
	Tome Rhodes at IW, but I don't think he wrote it...
*/	

extern sendbitmap();	/* encodes bitmap for output */
extern newsendbitmap();	/* encodes bitmap for output using new format */
extern readbitmap();	/* decodes bitmap from input */
extern printbmap();	/* formats bitmap for output */
extern read_pfd();	/* reads pfd structure elements */
extern doUNIXwork();	/* forwards stdin (if present) to the printer port */
extern touchtype();	/* forwards keyboard input (if present) to printer */

char filename[NCHFILE];	/* name of the file for read/write */
char pipewcmnd[NCHPIPE];	/* pipeline for pipe write */
char prname[NCHFILE];	/* ascii file containing pfd */
struct Printerdefs pfd;	/* printer definition structure */

Rectangle kbdrect; 	/* one line display */
Point	kbdp;		/* cursor point in one line display */

Rectangle srects[] = {
	 152, 8, 192, 32, 	/* location of lf button */
	 200, 8, 240, 32,	/* location of ff button */
	 0,   0,  0,   0
};

#define MAXCNT	2000

Bitmap  blit;		/* the screen (in a form we can manipulate) */

Bitmap *bp, *obp;	/* the selected bitmap */
Rectangle rect;		/* the selected rectangle */
struct Proc *proc,*oproc;/* the process selected. (when layer selected.) */
Point refpt;		/* position of view. */

short	status; 	/* keeps state info for revid, stop/run, and stipple */
char 	mustfree = 0;	/* should we free bp when finished with it? */
char 	hold = 0;	/* hold	io flag */
char	flush = 0;	/* flush io flag */
char	makecopy = 1;	/* make copy first flag */
char 	autores = 0;	/* automatic resolution selection on. */
char	dump = 0;	/* generate hex dump of formatted bitmap. */
#ifdef DMD630
char	maplf = 1;	/* map lf to lf/cr */
#else
char	maplf = 0;
#endif
int	in = 0;		/* amount of inset */

#if ! defined(PAR) && ! defined(DMD630)
char 	flow = 1;	/* do we try to regulate flow? */
#endif

#ifdef DMD630
extern cflag;
#endif

/* refresh display when laststate != state. */
char 	laststate = RES;/* the last state. */
char	state = SEL; 	/* the state we are in */	

char	vers;		/* which version terminal ROMS? */

main(argc, argv)
int	argc;
char	*argv[];
{
	int cnt;

	/* a version of the screen that we can mess with. */
	blit.base=addr(&display,Pt(0,0));
	blit.width=display.width;
	blit.rect=Jrect;

	vers = version() & 0x000000ff;
	request(RCV|KBD|MOUSE);
	parse_args(argc,argv);
#ifdef DMD630
	if (cflag)
		cache("dmdps",A_NO_SHOW);
#endif
	redraw(); 		

	sleep(120); /* jx on ibm with DACU FEP won't work without this.. */
	formp(INULL,INULL,INULL,"reading pfd file...");
	if (read_pfd(&pfd,prname)==-1){
		formp(INULL,INULL,INULL,"can't read pfd file.");
		sleep(60);
	}
	thinkstart();	/* initalize printer driver */
	formp(INULL,INULL,INULL,"turn on printer.");
	thinkflush();
	selnewbp();	/* get ready to select a new bitmap */

    	for (;wait(CPU);) {
		Reshape(); /* redraws only if necessary. */
		if (doUNIXwork())	/* download from dmdpr */
			cnt=0;	
		/* reset flushing download flag after download has stopped. */
		if (flush == TRUE && cnt++ < MAXCNT ) {
 			formp(INULL,INULL,INULL,"stopped flushing"); 
			alert();
			laststate = RES;
			flush=FALSE;
			cnt=0;
		}
		touchtype();	/* send kbd input to printer */
		thinkflush();

		if (state == SEL) {
			if (laststate != SEL) { 
				laststate = SEL;
 				formp(INULL,&menu2,&menu3,DMDPSVERS); 
				showbitmap(bp,inset(rect,in));
			}
			if (button3() && own()&MOUSE)
				do_bttn3();
			else if (bttn1() && own()&MOUSE)  
				do_bttn1();
			else if (bttn2() && own()&MOUSE)
				do_misc();
			if (bp) {
				state=OP;
				bttn3_menu[MSELOP] = "";
				oproc=proc;
				obp=bp;
				if (makecopy && !mustfree)
					clone_bitmap();
			}
		}
		if (state==OP) {
			if (laststate != OP) { 
			    	laststate=OP ;
			    	formp(&cross,&menu2,&menu3,"operations");
			    	showbitmap(bp,inset(rect,in));
			}
			if (button3() && own()&MOUSE) 
				do_bttn3();
			else if (bttn2() && own()&MOUSE)
				do_bttn2();
			else if (bttn1() && own()&MOUSE) 
				do_bttn1();
		}
	} /* main loop */
}

selnewbp()
{
	clearlayer();
	if (bp) {
		restore();
		if (mustfree) {
			bfree(bp);
			mustfree=0;
		}
		laststate=RES;
	}
	bp = BNULL; 
	proc = PNULL; 
	status= 0x00;
	state=SEL; 
	bttn3_menu[MSELOP] = NULL;
}

restore()
{
	if ( status&STIPPLE) {
		texture(bp,rect,&stiptext,F_XOR);
		status ^= STIPPLE;
	}
	if ( status&REVVID) {
		rectf(bp,rect,F_XOR);
		status ^= REVVID;
	}
	if ( status&STOPPED) {
		proc->state &= ~HALTED;
		status &= ~STOPPED;
	}
	in = 0; 				/* set inset to 0 */
}

readbp(in)
FILE *in;
{
	switch  (readbitmap(in)){
	case 0: formp(INULL,INULL,INULL, "no room");
		alert();
		break;
	case EOF:formp(INULL,INULL,INULL, "read failed");
		alert();
		break ;
	case 2: formp(INULL,INULL,INULL, "operation aborted");
		mustfree=2;
		break ;
	case 1: formp(INULL,INULL,INULL, "read complete");
		mustfree=2;
		return;
	}
	selnewbp();
}

doio(cmd,name)
int	cmd;
char	*name;
{
	FILE *filep;
	int rc;

	buttons(UP);
	formp(&thumbsdown,&thumbsdown,&thumbsup,"");
	laststate=RES;
	drstore(kbdrect);
	switch (cmd) {
	case WRITEPIPE:
	case READPIPE:
		kbdp = drstring("| ",LINE(kbdrect));
		break;
	case READFILE:
		kbdp = drstring("< ",LINE(kbdrect));
		break;
	case WRITEFILE:
		kbdp = drstring("> ",LINE(kbdrect));
		break;
	}
	kbdp=drstring(name,kbdp);
	if (!kbdstring(name,132,kbdp)) {
		formp(INULL,INULL,INULL, "operation aborted");
		sleep(30); laststate=RES; return;
	}
	cursswitch(&deadmouse);
	switch (cmd) {
	case READFILE:
		formp(INULL,&menu2,INULL, "reading from UN*X...");
		if ( (filep = fopen(name, "r")) == (FILE *) NULL) {
			formp(INULL,INULL,INULL, "open failed");
			selnewbp();
			alert();
			break;
		} 
		readbp(filep);
		break;
	case WRITEPIPE:
	case WRITEFILE:
		formp(INULL,&menu2,INULL, "sending to UN*X...");
		filep = ((cmd == WRITEPIPE) ? popen(name, "w" ) 
					    : fopen(name, "w"));
		if (filep == (FILE *) NULL) {
			formp(INULL,INULL,INULL, "couldn't create pipe/file");
			alert();
		} else {
			if ( newsendbitmap(bp,inset(rect,in),filep) == 1)
				formp(INULL,INULL,INULL, "output complete");
			else 
				formp(INULL,INULL,INULL, "output aborted");
		}
	}
	if (filep != (FILE *) NULL) {
		fflush(filep); 
		fclose(filep);
	}
 	laststate=RES; sleep(30);
	cursswitch(TNULL);
}

print (place,str)
Rectangle	place;
char *str;
{
	drstore(place);
	drstring(str, LINE(place));
	sleep(5);
}

formp(icon1, icon2, icon3, str) 
Texture16 *icon1;
Texture16 *icon2;
Texture16 *icon3;
char *str;
{
	fill_mouseboard(icon1, icon2, icon3);
	show_mouseboard(Pt(0,0));
	print(kbdrect,str);
}

redraw()
{
	int tx,ty;

	jrectf(Jrect,F_CLR); 
	tx=Drect.origin.x+(Drect.corner.x-Drect.origin.x)/2;
	ty=Drect.origin.y+(Drect.corner.y-Drect.origin.y)/2;
	if ( ty < Drect.origin.y+8) 
		ty=Drect.origin.y+8;
	else if ( ty > Drect.corner.y-8) 
		ty=Drect.corner.y-8;
	if ( tx < Drect.origin.x+8) 
		tx=Drect.origin.x+8;
	else if ( tx > Drect.corner.x-8) 
		tx=Drect.corner.x-8;
	refpt.y=ty; 
	refpt.x=tx;
	kbdrect=Drect;
	kbdrect.origin.y=kbdrect.corner.y-defont.height-4;
	mk_mouseboard(); 
	if (vers<5 && !(own()&MOUSE))
		texture(P->layer,Jrect,&stiptext,F_XOR);
	P->state &= ~RESHAPED;
	laststate=RES;
}

touchtype()
{
	char s[150];
	
	s[0]='\0';
	if ( own()&KBD ) {
#ifdef DMD630
		request(RCV|KBD|MOUSE|PSEND);
#endif
		formp(&thumbsdown,&thumbsdown,&thumbsup,"");
		laststate=RES;
		drstore(kbdrect);
		if ( kbdstring(s,150,LINE(kbdrect))) {
			shipstr(s); 
			if (maplf)
				shipchar('\r');	
			shipchar('\n');	
		}
#ifdef DMD630
		request(RCV|KBD|MOUSE);
#endif
	}
}
	
lexit3()	/* return true if button3 is clicked */
{
	Texture16 *prev; 
	int lexit;

	prev = (Texture16 *)cursswitch((Texture16 *) &thumbsdown);
	lexit = buttons(DOWN);buttons(UP);
	cursswitch(prev);
	return(lexit == 3);
}

buttons(updown)
int updown;
{
	do wait(MOUSE);
	while((bttn123()!=0) != updown || !ptinrect(mouse.xy,Drect));

	switch (bttn123()) {
		case 4: return 1;
		case 2: return 2;
		case 1: return 3;
	}
	return 0;
}

#define echo(ich,p)	bitblt(defont.bits,\
				Rect(ich->x,0,(ich+1)->x,defont.height),\
				&display,Pt(p.x+ich->left,p.y),F_XOR)

#define kbdcurs(p)	rectf(&display,Rect(p.x,p.y,p.x+1,p.y+defont.height),\
				F_XOR)

kbdstring(str,nchmax,p)	/* read string from keyboard with echo at p */
char *str; int nchmax; Point p; /* you pass a starting string in to edit. */
{
	int kbd, nchars; 
	Fontchar *ich;

	nchars=strlen(str); kbdcurs(p);
	for (;;) {
		sleep(3);
		if ( own()&KBD) {
			kbdcurs(p);
			do switch (kbd=kbdchar()) {
			case '\0':
				break;
			case '\r':
			case '\n':
				return nchars;
			case '\b':
				if (nchars <= 0) break;
				kbd=str[--nchars]; str[nchars]=0;
				ich=defont.info+kbd;
				p.x -= ich->width;
				echo(ich,p);
				break;
			default:
				if (nchars >= nchmax) break;
				str[nchars++]=kbd; str[nchars]=0;
				ich=defont.info+kbd;
				echo(ich,p);
				p.x += ich->width;
			} while (own()&KBD);
			kbdcurs(p);
		} else if ( own()&MOUSE && bttn12()) {
			buttons(UP);
			return 0;
		} else if ( own()&MOUSE && button3()) {
			buttons(UP);
			return nchars;
		}
	}
}

doprint(pfd)
struct Printerdefs *pfd;
{
	int returncode;

	cursswitch(&deadmouse);
	formp(INULL,&menu2,INULL, "printing...");
	if ((returncode = printbmap(pfd, bp, inset(rect,in))) == 1) 
		formp(INULL,INULL,INULL, "printing complete");
	else if ( returncode == 0 )
		formp(INULL,INULL,INULL, "printing aborted");
	else if ( returncode == -1 ) {
		formp(INULL,INULL,INULL, "unknown printer type");
		alert();
	}

	laststate=RES; sleep(20);
	cursswitch(TNULL);
	return returncode;
}

sbuthit (rects)
Rectangle rects[];
{
	register i;
	Point mickey;
	Rectangle r;

	mickey=mouse.xy;
	for ( i=0; !eqrect(rects[i],Rect(0,0,0,0)); i++) {
		r.origin=add(Drect.origin,rects[i].origin);
		r.corner=add(Drect.origin,rects[i].corner);
		if ( ptinrect(mickey,r) ) {
			rectf(&display,r,F_XOR);
			buttons(UP);
			rectf(&display,r,F_XOR);
			return i;
		}
	}
	return -1;
}

iocntrlmenu()
{
	int mhit;

	iomenutext[FLUSH][0]= flush?'*':' ';
	iomenutext[HOLD][0]= hold?'*':' ';
	
	if ( own()&MOUSE && bttn2() )	
	switch ( mhit=menuhit(&iocontrol_menu, 2)) {
	case FLUSH: 
		flush=flush?0:1;
		break;
	case HOLD: 
		hold=hold?0:1;
		break;
	}
}

#ifndef PAR
#ifndef DMD630
poptsmenu()
{
	int mhit;
	extern pbits, pspeed;

	do {
		port_opts[M7][0]= (pbits==7)?'*':' ';
		port_opts[M8][0]= (pbits==8)?'*':' ';
		port_opts[M9600][0]= (pspeed==9600)?'*':' ';
		port_opts[M4800][0]= (pspeed==4800)?'*':' ';
		port_opts[M1200][0]= (pspeed==1200)?'*':' ';

		if ( own()&MOUSE && bttn2() )
			switch (mhit=menuhit(&portopts, 2)){
			case M7: pbits=7;
				break;
			case M8: pbits=8;
				break;
			case M9600: pspeed=9600;
				break;
			case M4800: pspeed=4800;
				break;
			case M1200: pspeed=1200;
				break;
			case FLOWTICKS:
			case FLOWCHARS:
				buttons(UP); 
				if ( buttons(DOWN) == 2 )
					flowmenus(mhit,TRUE);
				break;
			}
		else if ( own()&MOUSE && bttn13()) 
			break;
		wait(CPU);
	} while ( mhit != -1 );
}
#endif
#endif

#ifndef PAR
#ifndef DMD630
flowmenus(which_one)
int which_one;
{
	int i, mhit;

	do {
		for (i=0;i<=LASTPOWER;i++) {
			tickmenu[i][0]=(pfd.flowticks&powertwo[i])?'*':' ';
			charsmenu[i][0]=(pfd.flowchars&powertwo[i])?'*':' ';
		}
		if ( which_one ==FLOWTICKS && own()&MOUSE && bttn2() 
		&& ((mhit=menuhit(&tick_menu,2)) != -1 ))
			pfd.flowticks ^= powertwo[mhit];
		else if ( which_one==FLOWCHARS  && own()&MOUSE && bttn2()
		&& ((mhit=menuhit(&chars_menu,2)) != -1 ))
			pfd.flowchars ^= powertwo[mhit];
		else if ( own()&MOUSE && bttn13()) 
			break;
		wait(CPU);
	} while ( mhit != -1 );
}
#endif 
#endif 

do_misc()
{
	int mhit;
	do {
		misc_menu[MONE2ONE]= (autores==0)?MCLIP:MSCALE;
		misc_menu[MCOPY][0]= (makecopy==1)?'*':' ';
		misc_menu[MDUMP][0]= (dump==1)?'*':' ';
		misc_menu[MHOLD][0]= (hold==1)?'*':' ';
		misc_menu[MFLUSH][0]= (flush==1)?'*':' ';
		misc_menu[MMAP][0]= (maplf==1)?'*':' ';

		if ( own()&MOUSE && bttn2())
			switch ( mhit=menuhit(&miscmenu, 2) ) {
			case MONE2ONE:
				autores=autores?0:1;
				break;
			case MCOPY:
				makecopy=makecopy?0:1;
				break;
			case MDUMP:
				dump=dump?0:1;
				break;
			case MEXIT:
				mybye();
				return;
			case MMAP:
				maplf=maplf?0:1;
				break;
			case MHOLD:
				hold=hold?0:1;
				break;
#ifndef PAR
#ifndef DMD630
			case MPOPTS:
				poptsmenu();
				break;
#endif
#endif
			case MFLUSH:
				flush=flush?0:1;
				break;
			case MREADPFD:
				formp(&thumbsdown,&thumbsdown,&thumbsup, "");
				laststate=RES;
				drstore(kbdrect);
				kbdp=drstring(prname,LINE(kbdrect));
				if (kbdstring(prname,NCHFILE,kbdp)) {
				    formp(INULL,INULL,INULL,"reading pfd...");
				    if (read_pfd(&pfd,prname)==-1){
					formp(INULL,INULL,INULL, 
						"couldn't read pfd file");
					alert();
				    } else {
				    	formp(INULL,INULL,INULL, "pfd read ok");
					sleep(20);
				    }
				} 
				return;
			}
		else if ( own()&MOUSE && bttn13()) 
			break;
		wait(CPU);
	} while (mhit != -1);
}

misc()
{
	int mhit;
	buttons(UP);
	formp(INULL,&menu2,INULL, "set options...");
	laststate=RES;
	if ( buttons(DOWN) == 2 ) do {
		if ( own()&MOUSE && bttn2() ) {
			mhit=do_misc();
			formp(INULL,&menu2,INULL, "set options...");
		}
		if ( own()&MOUSE && bttn13() ) {
			buttons(UP);
			break;
		}
		wait(CPU);
	} while (mhit != -1);
}

mybye()
{
	formp(&thumbsup,&thumbsup,&thumbsdown, "quit?");
	if (lexit3()) {
		thinkflush();
		thinkstop();
		exit(0);
	}
	laststate=RES;
} 

do_bttn2()
{
	int mhit;

	bit_output[MPRINT]=NULL;
	if ( prname[0] != '\0' ) 
		bit_output[MPRINT]="print bitmap";

	switch (mhit=menuhit(&bitoutput,2)) {
	case MWRITE: 
	case MPIPE: 
		doio(mhit==MPIPE ?WRITEPIPE:WRITEFILE,
		     mhit==MPIPE ?pipewcmnd:filename);
		break;
	case MPRINT: 
#ifdef DMD630
		request(RCV|KBD|MOUSE|PSEND);
#endif
		doprint(&pfd);
#ifdef DMD630
		request(RCV|KBD|MOUSE);
#endif
		break;
	default: state=OP;
	}
}

do_bttn3()
{
	int mhit;

	switch (mhit = menuhit(&bttn3menu,3)) {
	case MSEL_LAYER:	/* choose layers */
		formp(INULL,INULL,&target, "select a layer");
		selnewbp();
		if ((proc=debug()) != PNULL){ 
			bp=(Bitmap *)proc->layer; 
			rect=bp->rect;
			flash(bp,rect);
		}  else
			selnewbp();
		cursswitch(TNULL); /* debug is hosed */
		break;
	case MSEL_RECT: /* sweep out a rectangle */
		formp(INULL,INULL,&sweepcursor, "sweep a rectangle");
		selnewbp();
		bp= &blit; 
		rect=getrect();
		if ((rect.corner.x-rect.origin.x)<XMIN
		&&  (rect.corner.y-rect.origin.y)<YMIN)
			selnewbp();
		else
			flash(bp,rect);
		break;
	case MSEL_SCREEN:	/* whole screen */
		selnewbp();
		bp= &blit; 
		rect=Jrect;
		flash(bp,rect);
		break;
	case MSEL_FILE: 
		selnewbp();
		doio(READFILE,filename);
		rect=bp->rect;
		break;
	case MINSET: 
		flash(bp,inset(rect,in += 4));
		clearlayer();
		break;
	case MFLIPSTIP:	
		texture(bp,rect,&stiptext,F_XOR);
		status ^= STIPPLE;
		break; /* op */
	case MFLIPVID:	
		rectf(bp,rect,F_XOR);
		status ^= REVVID;
		break; /* op */
	case MCOPYOP:
		clone_bitmap();
		break;
	case MRUNHALT: 
		if ( (proc->state)&HALTED ){
			proc->state &= ~HALTED;
			status &= ~STOPPED;
		} else {
			proc->state |= HALTED;
			status |= STOPPED;
		}
		break; 
	case MISC:
		misc();
		break;
	} /* end switch */
	laststate=RES;
}

do_bttn1()
{
#ifdef DMD630
	request(RCV|KBD|MOUSE|PSEND);
#endif
	switch (sbuthit(srects)){
	case 0: shipchar('\n');
#ifdef DMD630
		shipchar('\r');
#endif
		buttons(UP);
		break;
	case 1: shipchar('\f');
		buttons(UP);
		break;
	default:
		laststate=RES;	/* force a call to showbitmap */
		break;
	}
	thinkflush();
#ifdef DMD630
	request(RCV|KBD|MOUSE);
#endif
}

clone_bitmap()
{
	Bitmap *clonemap;

	clonemap = BNULL;
	/* should test to make sure proc is still there.... */
	if (oproc) 
		obp = (Bitmap *) oproc->layer;
	if (mustfree==2)
		formp(INULL,INULL,INULL, "using a copy...");
	else if ( obp && (clonemap = balloc(rect)) ) {
		bitblt(obp,inset(rect,in),clonemap,add(rect.origin,in),F_STORE);
		restore();
		if (mustfree)
			bfree(bp);
		bp=clonemap;
		mustfree=1;
		proc=PNULL;
		formp(INULL,INULL,INULL, "made a copy...");
	} else 
		formp(INULL,INULL,INULL, "using the screen image...");
	sleep(15);
	laststate = RES;
}

clearlayer()
{
	Rectangle dstrect;

	dstrect=P->layer->rect;
	dstrect.origin=add(dstrect.origin,Pt(8,38));
	dstrect.corner=add(dstrect.corner,Pt(-8,-(defont.height+8)));
	rectf(P->layer,dstrect,F_CLR);
}

alert()
{
	ringbell();
	flash(P->layer,P->layer->rect);
	sleep(20);
}

flash(b,r)
Bitmap *b;
Rectangle r;
{
	rectf(b,r,F_XOR);
	nap(20);
	rectf(b,r,F_XOR);
}
